examples/kv-server/src/main/java/com/yahoo/gondola/container/RoutingFilter.java

```diff
@@ -0,0 +1,157 @@
+
+    static Gondola gondola;
+
+    CloseableHttpClient httpclient;
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+        httpclient = HttpClients.createDefault();
+
+        String clusterCallbackClass = filterConfig.getInitParameter("clusterCallback");
+        if (clusterCallbackClass != null) {
+            registerClusterCallback(clusterCallbackClass);
+        }
+        registerRoutingTable();
+    }
+
+    // TODO: move to routing table impl
+    private void registerRoutingTable() {
+        routingTable.put(81, "http://localhost:8080");
+        routingTable.put(82, "http://localhost:8081");
+        routingTable.put(83, "http://localhost:8082");
+    }
+
+    // TODO: get server scheme://hostname:port
+    private String getServerURI() {
+        return "http://localhost:8080";
+    }
+
+
+    // TODO: discover
+    private void registerClusterCallback(String clusterCallbackClass) throws ServletException {
+        try {
+            this.callback = (ClusterCallback)RoutingFilter.class.getClassLoader()
+                .loadClass(clusterCallbackClass)
+                .newInstance();
+        } catch (ClassNotFoundException|InstantiationException|IllegalAccessException e) {
+            throw new ServletException(e);
+        }
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+        throws IOException, ServletException {
+        Member leader = getLeader(request);
+
+        if (isRemoteLeader(leader) && request instanceof HttpServletRequest) {
+            proxyRequestToLeader((HttpServletRequest) request, response, leader);
+        } else {
+            chain.doFilter(request, response);
+        }
+    }
+
+    private boolean isRemoteLeader(Member leader) {
+        return leader != null && !leader.isLocal();
+    }
+
+    private Member getLeader(ServletRequest request) {
+        if (gondola == null) {
+            return null;
+        }
+
+        return callback != null ? getLeaderByHelper(request) : getDefaultLeader();
+    }
+
+    private Member getLeaderByHelper(ServletRequest request) {
+        Member leader;
+        String clusterId = callback.getClusterId(request);
+        leader = gondola.getCluster(clusterId).getLeader();
+        return leader;
+    }
+
+    private Member getDefaultLeader() {
+        return gondola.getClustersOnHost().get(0).getLeader();
+    }
+
+    private void proxyRequestToLeader(HttpServletRequest request, ServletResponse response, Member leader)
+        throws IOException {
+        String destHost = routingTable.get(leader.getMemberId());
+        HttpServletResponse httpServletResponse = (HttpServletResponse) response;
+        String method = request.getMethod();
+        String requestURI = request.getRequestURI();
+        CloseableHttpResponse proxiedResponse;
+        if (method.equals("GET")) {
+            HttpGet httpGet = new HttpGet(destHost + requestURI);
```

Comment:
did look the apache httpclient request class, it requires to give specific HTTP method to send the request. There should be a library fit the forwarding use case better.